awx/main/tests/unit/utils/test_common.py

```diff
@@ -178,10 +178,21 @@ def test_get_custom_venv_choices():
 
     with TemporaryDirectory(dir=settings.BASE_VENV_PATH, prefix='tmp') as temp_dir:
         os.makedirs(os.path.join(temp_dir, 'bin', 'activate'))
+
+        custom_venv_dir = os.path.join(temp_dir, 'custom')
+        custom_venv_1 = os.path.join(custom_venv_dir, 'venv-1')
+        custom_venv_awx = os.path.join(custom_venv_dir, 'custom', 'awx')
+
+        os.makedirs(os.path.join(custom_venv_1, 'bin', 'activate'))
+        os.makedirs(os.path.join(custom_venv_awx, 'bin', 'activate'))
+
+        settings.CUSTOM_VENV_PATHS = [custom_venv_dir]
```

Comment:
One problem we regularly have with changing settings in this way is that doing this and _not_ resetting the original value can sometimes cause this state to leak between tests, and can have side effects when *other* tests run.  In really bad situations, it can cause tests failures that **only** occur when running all the tests, but tests pass individually 😱.

There are ways around this, like mocking `settings`, and a variety of decorators and context managers that can automatically _reset_ the originally value after this test runs, but they can be messy and easy to get wrong.

One pattern I really like for situations like this is to add an optional argument e.g.,

```python
def get_custom_venv_choices(custom_paths=None):
    custom_paths = custom_paths or settings.CUSTOM_VENV_PATHS
```

...and then have the *test* just call:

`common.get_custom_venv_choices([custom_venv_dir])`

Reply:
Yes, I believe that's a problem if we directly set the variables. And although, redefining the function is an option here, I am thinking of going with the mock approach. Because it is more elegant that way as we don't have to modify our code pertaining to tests, but instead, we modify tests pertaining to code.

```
with mock.patch('django.conf.settings.CUSTOM_VENV_PATHS', [custom_venv_dir]):
            assert sorted(common.get_custom_venv_choices()) == [
            ....
```

Reply:
> Because it is more elegant that way as we don't have to modify our code pertaining to tests, but instead, we modify tests pertaining to code.

There _is_ a distinction between modifying code to accommodate tests, and writing code that is _easily testable_.  The pattern I'm referring to above is known as dependency injection, and you'll see it used a lot throughout the unit tests in our codebase as an _alternative_ to mocking.

That said, much of this comes down to personal preference, and you'll find that every developer you talk to has a different opinion on when it is and isn't appropriate to use a mock 😄.

Reply:
I didn't know about that, thanks for pointing it out. Then I will go with this approach. :+1:

Reply:
The `django.test` module has a `modify_settings` decorator that may still work even though we aren't using Django's built-in testing classes here.

Reply:
Thanks @jbradberry. I thought of that at first, but since we are aiming to make it easy for future developers to test their code, I have for now, gone with the dependency injection approach. Let's discuss on this more, if we feel we can make some changes.